yt/yt/client/logging/dynamic_table_log_writer.cpp

```diff
@@ -0,0 +1,372 @@
+    : public virtual ILogWriter
+    , public TRateLimitingLogWriterBase
+    , public TSegmentSizeReporter
+{
+public:
+    TDynamicTableLogWriter(
+        TClientHolderPtr clientHolder,
+        std::unique_ptr<ILogFormatter> formatter,
+        const TDynamicTableLogWriterConfigPtr& config,
+        const TString& name,
+        IInvokerPtr invoker)
+        : TRateLimitingLogWriterBase(
+            BIND(&TDynamicTableLogWriter::WriteImpl, Unretained(this)),
+            name,
+            config)
+        , TSegmentSizeReporter(name)
+        , ClientHolder_(std::move(clientHolder))
+        , Formatter_(std::move(formatter))
+        , Config_(config)
+        , Invoker_(std::move(invoker))
+    , FlushExecutor_(New<TPeriodicExecutor>(
+            Invoker_,
+            BIND(&TDynamicTableLogWriter::DoFlush, MakeWeak(this)),
+            Config_->FlushPeriod))
+        , Logger(SystemLogger().WithTag("LogWriterName: %v, TablePath: %v", name, Config_->TablePath))
+    {
+        FlushExecutor_->Start();
+
+        YT_LOG_INFO("Created dynamic table log writer (Config: %v)", ConvertToYsonString(Config_, EYsonFormat::Text));
+    }
+
+    // No synchronous YT writes/calls are actually performed by Write/Flush, since we cannot
+    // switch contexts in log manager's main logging thread.
+    // Moreover, the current log manager is not designed for handling asynchronous event
+    // writers. All write/flush calls are made consecutively and are considered successful
+    // at once. Changing this would require modifying the logging event pipeline significantly.
+    i64 WriteImpl(const TLogEvent& event)
+    {
+        VERIFY_THREAD_AFFINITY(LoggingThread);
+
+        // This writer has its own buffer, so we reimplement some of the log manager logic :(
+        auto backlogWeight = BacklogWeight_.load(std::memory_order::relaxed);
+
+        if (Suspended_ && backlogWeight < Config_->LowBacklogWeightWatermark) {
+            Suspended_ = false;
+            YT_LOG_INFO("Backlog weight has dropped below low watermark, dynamic table logging resumed (LowBacklogWeightWatermark: %v)",
+                Config_->LowBacklogWeightWatermark);
+        }
+
+        if (!Suspended_ && backlogWeight > Config_->HighBacklogWeightWatermark) {
+            Suspended_ = true;
+            YT_LOG_WARNING("Backlog weight has exceeded high watermark, dynamic table logging suspended (HighBacklogWeightWatermark: %v)",
+                Config_->HighBacklogWeightWatermark);
+        }
+
+        if (Suspended_) {
+            return 0;
+        }
+
+        auto guard = Guard(SpinLock_);
+
+        // This is necessary to avoid hitting the limit of allowed row modifications within a single transaction.
+        // Also, there is no upside from huge batches.
+        if (CurrentRowCount_ > Config_->MaxBatchRowCount || std::ssize(CurrentBuffer_) > Config_->MaxBatchWeight) {
+            RotateBuffer();
+        }
+
+        TStringOutput outputStream(CurrentBuffer_);
```

Comment:
Why do we convert batch of events to a single string first and then convert them into a list of rows? Is it possible to create rows here?

Reply:
The main problem is that we need to know the table schema in order to convert the yson strings into unversioned rows, and getting the schema is a YT client request.

The tricky part is that the schema might change, for example, we forgot to add a column, saw the errors in logs and then altered the table to contain the forgotten column.

So either we need some periodic that will fetch and store the schema, or we fetch it upon write.

For now I decided to go with the second approach. Later on, if we switch to some writer that needs to know the schema upon creation, we can switch to the first approach and move parsing into `Write`. However, the writer will have to support schema-alerting :)

Reply:
This is considered resolved (without any changes), but I will keep this thread open for now, so that future reviewers see it.